### DISCUSSION TITLE
4.38-beta3 : WebAuthn failed : "userHandle and User ID do not match"

### DISCUSSION THREAD

**Discussion by lbalogh** [authelia/authelia#6465] [Pre-Release Feedback]
Hi,

I upgraded Authelia from v4.37.5 to v4.38.0-beta3 by changing the tag in my Docker compose file ("latest" -> "v4.38.0-beta3"), and since then WebAuthn MFA doesn't work : I get a "Failed to initiate security key sign in process" popup. The logs show :

`
time="2023-12-27T09:29:47+01:00" level=error msg="Unsuccessful WebAuthn authentication attempt by user 'xxx_admin'" error="userHandle and User ID do not match" method=POST path=/api/secondfactor/webauthn remote_ip=192.168.1.1 stack="github.com/authelia/authelia/v4/internal/handlers/response.go:275              markAuthenticationAttempt\ngithub.com/authelia/authelia/v4/internal/handlers/handler_sign_webauthn.go:193 WebAuthnAssertionPOST\ngithub.com/authelia/authelia/v4/internal/middlewares/require_auth.go:19        Require1FA.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/bridge.go:54              handleRouter.(*BridgeBuilder).Build.func6.1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:56             SecurityHeadersCSPNone.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:82             SecurityHeadersNoStore.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:12             SecurityHeaders.func1\ngithub.com/fasthttp/router@v1.4.22/router.go:420                               (*Router).Handler\ngithub.com/authelia/authelia/v4/internal/middlewares/log_request.go:14         handleRouter.LogRequest.func40\ngithub.com/valyala/fasthttp@v1.51.0/server.go:2359                             (*Server).serveConn\ngithub.com/valyala/fasthttp@v1.51.0/workerpool.go:224                          (*workerPool).workerFunc\ngithub.com/valyala/fasthttp@v1.51.0/workerpool.go:196                          (*workerPool).getCh.func1\nruntime/asm_amd64.s:1650                                                       goexit"
`

I used my iPhone as security key, which is synced to my Mac through iCloud. Whenever Authelia requests the key, a popup appears on my Mac prompting me to unlock it using Touch ID, but then, I encounter the mentioned error.

It was working fine in v4.37.5.

Build info :

```
# authelia build-info
Last Tag: v4.37.5
State: untagged clean
Branch: v4.38.0-beta3
Commit: e26aa48a2aedadd834d44e333a2958c8c35ba0ef
Build Number: 25882
Build OS: linux
Build Arch: amd64
Build Date: Tue, 26 Dec 2023 16:30:38 +1100
```

Any idea ?

**Comment by james-d-elliott** [comment]
Can you confirm deleting and adding a new credential works and fixes the error?

**Comment by lbalogh** [comment]
So I deleted my current key from iCloud keychain (iPhone + Mac) and from Authelia, and added a new one. Now the error is different :

`time="2023-12-27T10:58:18+01:00" level=error msg="Error occurred validating a WebAuthn authentication challenge for user 'xxx_admin': error parsing the request body" error="Parse error for Assertion (invalid_request): illegal base64 data at input byte 7" method=POST path=/api/secondfactor/webauthn remote_ip=192.168.1.1 stack="github.com/authelia/authelia/v4/internal/handlers/handler_sign_webauthn.go:157 WebAuthnAssertionPOST\ngithub.com/authelia/authelia/v4/internal/middlewares/require_auth.go:19        Require1FA.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/bridge.go:54              handleRouter.(*BridgeBuilder).Build.func6.1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:56             SecurityHeadersCSPNone.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:82             SecurityHeadersNoStore.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:12             SecurityHeaders.func1\ngithub.com/fasthttp/router@v1.4.22/router.go:420                               (*Router).Handler\ngithub.com/authelia/authelia/v4/internal/middlewares/log_request.go:14         handleRouter.LogRequest.func40\ngithub.com/valyala/fasthttp@v1.51.0/server.go:2359                             (*Server).serveConn\ngithub.com/valyala/fasthttp@v1.51.0/workerpool.go:224                          (*workerPool).workerFunc\ngithub.com/valyala/fasthttp@v1.51.0/workerpool.go:196                          (*workerPool).getCh.func1\nruntime/asm_amd64.s:1650                                                       goexit"`

**Reply by james-d-elliott** [reply]
Interesting, I can't seem to replicate either issue with a YubiKey. Wonder if it's a browser specific or device specific nuance.

**Reply by james-d-elliott** [reply]
Looks like it is mac specific: #6441

I don't have a mac to test with so it will be interesting trying to figure out what this is I guess.

**Reply by james-d-elliott** [reply]
I'll create a build which should give more information, but just trying to think of all the potential scenarios that may cause this now.

**Reply by james-d-elliott** [reply]
Can you show the difference with the `v4.38.0-beta3-wd` build?

**Comment by lbalogh** [comment]
Thank you for taking the time to look after that. Here is the log using the v4.38.0-beta3-wd build :

`time="2023-12-27T13:51:00+01:00" level=error msg="Error occurred validating a WebAuthn authentication challenge for user 'xxx_admin': error parsing the request body" error="Parse error for Assertion (invalid_request): Error reading '{\"id\":\"mz-LQB5VYSF_WsLXPL4ba9rFEv0\",\"rawId\":\"mz-LQB5VYSF_WsLXPL4ba9rFEv0\",\"response\":{\"authenticatorData\":\"nkStRPzSYv9P4zc0rHkZS1KuGyfJKgabGOWkWnjOvI4dAAAAAA\",\"clientDataJSON\":\"eyJ0eXBlIjoid2ViYXV0aG4uZ2V0IiwiY2hhbGxlbmdlIjoiVW5sR3pVYUpNVzFWaVBQUlozQWlfQktjQUY5cnVBSHZvYnplczBBN0xORSIsIm9yaWdpbiI6Imh0dHBzOi8vYXV0aC5sYmFsb2doLmNoIiwiY3Jvc3NPcmlnaW4iOmZhbHNlfQ\",\"signature\":\"MEUCIE5nipQCHDt5iyN3ZeSFM6PxfAij0LieadovGuKiiVI9AiEA-JYBEdmB3HXlsag5J1rEXfZNOlAFdSJByIsLuZnUBJk\",\"userHandle\":\"cMluyRc?#l@eXE9V5jqj{ceG2+GoWpQG}nXS1N\\\"'\\\\Ik0yzg2&K%wutDt@HS\\\"$O/}\"},\"type\":\"public-key\",\"clientExtensionResults\":{},\"authenticatorAttachment\":\"platform\"}': failed to decode value 'cMluyRc?#l@eXE9V5jqj{ceG2+GoWpQG}nXS1N\\\"'\\\\Ik0yzg2&K%wutDt@HS\\\"$O/}' as base64: illegal base64 data at input byte 7" method=POST path=/api/secondfactor/webauthn remote_ip=192.168.1.1 stack="github.com/authelia/authelia/v4/internal/handlers/handler_sign_webauthn.go:157 WebAuthnAssertionPOST\ngithub.com/authelia/authelia/v4/internal/middlewares/require_auth.go:19        Require1FA.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/bridge.go:54              handleRouter.(*BridgeBuilder).Build.func6.1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:56             SecurityHeadersCSPNone.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:82             SecurityHeadersNoStore.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:12             SecurityHeaders.func1\ngithub.com/fasthttp/router@v1.4.22/router.go:420                               (*Router).Handler\ngithub.com/authelia/authelia/v4/internal/middlewares/log_request.go:14         handleRouter.LogRequest.func40\ngithub.com/valyala/fasthttp@v1.51.0/server.go:2359                             (*Server).serveConn\ngithub.com/valyala/fasthttp@v1.51.0/workerpool.go:224                          (*workerPool).workerFunc\ngithub.com/valyala/fasthttp@v1.51.0/workerpool.go:196                          (*workerPool).getCh.func1\nruntime/asm_amd64.s:1650`

**Reply by james-d-elliott** [reply]
Huh.. interesting. Time to do some research. Seems Apple is returning the `userHandle` different to everyone else maybe, otherwise just really unlucky.. what's really interesting is you're able to register it since we specifically request `cross-platform` but it's a `platform` credential.

**Comment by lbalogh** [comment]
Let me know if you need me to do some more testing.

I just tested again in other browsers (Chrome and Safari on Mac, Chrome on Windows), using the same key generated on my iPhone, and they all fail the same way

**Reply by james-d-elliott** [reply]
Yeah I think I'll need to. Think it's a bug with a library I maintain but no one else has had it for some inexplicable reason. But it's used by a lot of repositories so I have a pretty large chance of breaking their stuff as well so probably going to have to do a workaround and plan a breaking change.

**Reply by james-d-elliott** [reply]
Give it another shot after pulling again when you get time.

**Reply by lbalogh** [reply]
Here is the log with the latest pull :

`time="2023-12-27T15:57:25+01:00" level=error msg="Error occurred validating a WebAuthn authentication challenge for user 'xxx_admin': error parsing the request body" error="Parse error for Assertion (invalid_request): Error reading '{\"id\":\"mz-LQB5VYSF_WsLXPL4ba9rFEv0\",\"rawId\":\"mz-LQB5VYSF_WsLXPL4ba9rFEv0\",\"response\":{\"authenticatorData\":\"nkStRPzSYv9P4zc0rHkZS1KuGyfJKgabGOWkWnjOvI4dAAAAAA\",\"clientDataJSON\":\"eyJ0eXBlIjoid2ViYXV0aG4uZ2V0IiwiY2hhbGxlbmdlIjoidFZnWFFqOGhEejBFZDg1WWExcVZPQTNZUGl5ZFY3VU9JUkFpVjN1VU1LNCIsIm9yaWdpbiI6Imh0dHBzOi8vYXV0aC5sYmFsb2doLmNoIiwiY3Jvc3NPcmlnaW4iOmZhbHNlfQ\",\"signature\":\"MEQCIAsO9soBdUu7Jo_bje2_N_49tIbjMTCEH82TlglSuWL6AiAbBGNwf_srN9VqnNgINGHDa4QW3GAGgoCNrCULNW6fkQ\",\"userHandle\":\"cMluyRc?#l@eXE9V5jqj{ceG2+GoWpQG}nXS1N\\\"'\\\\Ik0yzg2&K%wutDt@HS\\\"$O/}\"},\"type\":\"public-key\",\"clientExtensionResults\":{},\"authenticatorAttachment\":\"platform\"}': failed to decode value 'cMluyRc?#l@eXE9V5jqj{ceG2+GoWpQG}nXS1N\\\"'\\\\Ik0yzg2&K%wutDt@HS\\\"$O/}' as base64: illegal base64 data at input byte 7" method=POST path=/api/secondfactor/webauthn remote_ip=192.168.1.1 stack="github.com/authelia/authelia/v4/internal/handlers/handler_sign_webauthn.go:157 WebAuthnAssertionPOST\ngithub.com/authelia/authelia/v4/internal/middlewares/require_auth.go:19        Require1FA.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/bridge.go:54              handleRouter.(*BridgeBuilder).Build.func6.1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:56             SecurityHeadersCSPNone.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:82             SecurityHeadersNoStore.func1\ngithub.com/authelia/authelia/v4/internal/middlewares/headers.go:12             SecurityHeaders.func1\ngithub.com/fasthttp/router@v1.4.22/router.go:420                               (*Router).Handler\ngithub.com/authelia/authelia/v4/internal/middlewares/log_request.go:14         handleRouter.LogRequest.func40\ngithub.com/valyala/fasthttp@v1.51.0/server.go:2359                             (*Server).serveConn\ngithub.com/valyala/fasthttp@v1.51.0/workerpool.go:224                          (*workerPool).workerFunc\ngithub.com/valyala/fasthttp@v1.51.0/workerpool.go:196                          (*workerPool).getCh.func1\nruntime/asm_amd64.s:1650`

Build info :

```# authelia build-info
Last Tag: v4.37.5
State: untagged clean
Branch: v4.38.0-beta3-wd
Commit: f5a56fc01ae9f8765e3d7618a462caf7713d1608
Build Number: 25899
Build OS: linux
Build Arch: amd64
Build Date: Thu, 28 Dec 2023 00:45:07 +1100```

**Reply by james-d-elliott** [reply]
Did you recreate the credential by chance?

**Reply by lbalogh** [reply]
I just tried to deleted and re add my key (still on build number 25899) but this time I can't. I get this screen : 

![image](https://github.com/authelia/authelia/assets/6897141/1652224e-53a5-495a-94c8-311fee688744)

And after clicking on Next :

![image](https://github.com/authelia/authelia/assets/6897141/e1a49bba-ee7a-4af6-a148-09bcb68ec631)

Nothing in the logs.

After that I revert back to tag v4.38.0-beta3, and I could register my key just fine (except login still doesn't work because of the "error parsing the request body..." we are discussing)

**Reply by james-d-elliott** [reply]
I believe I fixed it on the original beta.

**Reply by lbalogh** [reply]
I just tried again with the updated original beta, and it worked ! I could log in using my Apple device as security key. Thanks

```# authelia build-info
Last Tag: v4.37.5
State: untagged clean
Branch: v4.38.0-beta3
Commit: 871201696950a3e4b2ac2d0c1f7aa492c2e93976
Build Number: 25995
Build OS: linux
Build Arch: amd64
Build Date: Fri, 29 Dec 2023 16:34:51 +1100```

**Reply by james-d-elliott** [reply]
Nice!
